cfg/builder/builder_finalize.cc

```diff
@@ -130,92 +131,102 @@ void CFGBuilder::sanityCheck(core::Context ctx, CFG &cfg) {
         auto &bb = *it;
         if (bb == cfg.deadBlock()) {
             continue;
         }
-        UnorderedMap<core::LocalVariable, core::LocalVariable> &current = outAliases[bb->id];
+        auto &current = outAliases[bb->id];
+
         if (!bb->backEdges.empty()) {
+            // Take the intersection of all of the back edges' aliases.
             current = outAliases[bb->backEdges[0]->id];
-        }
-
-        for (BasicBlock *parent : bb->backEdges) {
-            UnorderedMap<core::LocalVariable, core::LocalVariable> other = outAliases[parent->id];
-            for (auto it = current.begin(); it != current.end(); /* nothing */) {
-                auto &el = *it;
-                auto fnd = other.find(el.first);
-                if (fnd != other.end()) {
-                    if (fnd->second != el.second) {
-                        current.erase(it++);
+            for (auto i = 1; i < bb->backEdges.size(); i++) {
+                auto *parent = bb->backEdges[i];
+                const auto &other = outAliases[parent->id];
+                for (auto it = current.begin(); it != current.end(); /* nothing */) {
+                    auto &el = *it;
+                    auto fnd = other.find(el.first);
+                    if (fnd != other.end()) {
+                        if (fnd->second != el.second) {
+                            current.erase(it++);
+                        } else {
+                            ++it;
+                        }
                     } else {
-                        ++it;
+                        current.erase(
+                            it++); // note: this is correct but too conservative. In particular for loop headers
                     }
-                } else {
-                    current.erase(it++); // note: this is correct but to conservative. In particular for loop headers
                 }
             }
         }
 
+        // Overapproximation of the set of variables that have aliases.
+        // Will have false positives, but no false negatives. Avoids an expensive inner loop below.
+        UnorderedSet<LocalRef> mayHaveAlias;
+        for (auto &alias : current) {
+            mayHaveAlias.insert(alias.second);
+        }
+
         for (Binding &bind : bb->exprs) {
             if (auto *i = cast_instruction<Ident>(bind.value.get())) {
-                i->what = maybeDealias(ctx, i->what, current);
+                i->what = maybeDealias(ctx, cfg, i->what, current);
             }
-            /* invalidate a stale record */
-            for (auto it = current.begin(); it != current.end(); /* nothing */) {
-                if (it->second == bind.bind.variable) {
-                    current.erase(it++);
-                } else {
-                    ++it;
+            if (mayHaveAlias.contains(bind.bind.variable)) {
+                /* invalidate a stale record (uncommon) */
+                for (auto it = current.begin(); it != current.end(); /* nothing */) {
```

Comment:
Why not increment the iterator here instead? The combination of pre and post-increment in the loop body is a little hard to follow.

Reply:
`erase` mutates the iterator.

But actually, you raise a good point. The code should probably do:

```
it = current.erase(it);
```

rathe than

```
current.erase(it++);
```

Reply:
Ah good point. Iterator api strikes again.

Reply:
Turns out the `current.erase(it++)` is by necessity.